=doc #18657 Doc java directives

patriknw opened the pull request.

patriknw committed:
=doc #18657 Doc host and path directives

* also including docs of RequestVals.matchAndExtractHost
  and RequestVals.matchAndExtractHost
* and add Host.create factory methods
* add missing HttpRequest PATCH and OPTIONS
* change to val in matchAndExtractHost for
  fail fast exception

patriknw commented on akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java:
```diff
@@ -0,0 +1,34 @@
+    final Route shortOnly = host(hostname -> hostname.length() < 10,
+        completeWithStatus(StatusCodes.OK));
+
+    path("test").route(shortOnly);
```
Here is corresponding scala documentation: http://doc.akka.io/docs/akka-stream-and-http-experimental/snapshot/scala/http/routing-dsl/directives/host-directives/host.html#host

I don't know if this path("test") makes sense?

The Scala documentation shows test code: `Get() ~> Host("rest.company.com") ~> route ~> check`

ktoso replied:
I think what we should show here is the javadsl testkit:

```java
        // test responses to potential errors
        appRoute.run(HttpRequest.GET("/")) // TODO make it be to host api.company.com
            .assertStatusCode(StatusCodes.OK);
```

etc

See here: https://github.com/akka/akka/blob/release-2.3-dev/akka-docs-dev/rst/java/code/docs/http/javadsl/server/testkit/TestkitExampleTest.java

drewhk commented on akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java:
```diff
@@ -0,0 +1,34 @@
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package docs.http.javadsl.server.directives;
```
how come this file is not in akka-samples/akka-docs-java-lambda? We don't need Java8 stuff?

ktoso replied:
We're going ahead with java8 in docs here.

drewhk replied:
that is great. Why can't we do the same for streams?

ktoso replied:
I thought we are? The cookbook used java8

ktoso commented on akka-http-core/src/main/java/akka/http/javadsl/model/headers/Host.java:
```diff
@@ -4,7 +4,22 @@
+    
+    public static Host create(String host, int port) {
+      return akka.http.scaladsl.model.headers.Host.apply(host, port);
+    }
```
Great catch, thanks @patriknw !

patriknw commented on akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/HostDirectivesExamplesTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    testRoute(matchListOfHosts).run(
+        HttpRequest.GET("/").addHeader(Host.create("api.company.com")))
+        .assertStatusCode(StatusCodes.OK);
```
This does not work, and I can't find what is wrong. To me it looks like the host is extracted from the uri and not from the header (as it should according to the documentation), but I can't see why the Scala test works.

patriknw commented:
I have added `RequestVal.host` (extractHost) and all method directives.

patriknw commented:
I'll add the `RequestVal.matchAndExtractHost` also. It's the regexp variant of `host`.

patriknw commented on akka-docs-dev/rst/java/code/docs/http/javadsl/server/directives/MethodDirectivesExamplesTest.java:
```diff
@@ -0,0 +1,129 @@
+  //    testRoute(route).run(HttpRequest.PATCH("/", "patch content")).assertEntity(
+  //        "This is a PATCH request.");
+  //    //#patch
+  //  }
```
does anyone know?

ktoso replied:
The request method itself is `akka.http.javadsl.model.HttpMethods.PATCH` etc.  @patriknw

ktoso replied:
Seems like another catch actually, should be also made available on `akka.http.javadsl.model.HttpRequest` as the other methods for ease of testing.

// cc @sirthias @jrudolph

ktoso replied:
i.e.

```
public static HttpRequest PATCH(String uri) {
  return create(uri).withMethod(HttpMethods.PATCH);
}
```

Can you add it please @patriknw ?

jrudolph commented:
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3824/:

```
 16288 (akka-docs-dev/sphinx:generatePdf) Sphinx pdf generation failed.  See debug output for details.
 16289 Total time: 779 s, completed Oct 7, 2015 4:35:38 PM
```

patriknw commented:
I have added docs for `RequestVals.matchAndExtractHost` and added the missing PATCH and OPTIONS.

patriknw commented:
Regarding the sphinx error I can reproduce it locally, but I have no clue what is causing it.

ktoso commented on akka-docs-dev/rst/java/http/routing-dsl/directives/host-directives/extractHost.rst:
```diff
@@ -0,0 +1,13 @@
+.. _-extractHost-java-:
```
We'll need to figure out where to document request values so it's consistent.. For now looks good :)

ktoso commented on akka-docs-dev/rst/java/http/routing-dsl/directives/method-directives/index.rst:
```diff
@@ -0,0 +1,18 @@
+   head
+   method
+   options
+   overrideMethodWithParameter
```
Put this one last?

ktoso commented:
LGTM, will merge once the sphinx fix is merged and this can be validated then

ktoso commented:
PLS BUILD

patriknw commented:
I'll squash first

ktoso commented:
Great thanks

patriknw commented:
Squashed and rebased. This PR includes:

=doc #18657 Doc host and path directives

* also including docs of RequestVals.matchAndExtractHost
  and RequestVals.matchAndExtractHost
* and add Host.create factory methods
* add missing HttpRequest PATCH and OPTIONS
* change to val in matchAndExtractHost for
  fail fast exception

patriknw commented:
Refs #18657
